keras/engine/training.py

```diff
@@ -258,7 +259,14 @@ def collect_trainable_weights(layer):
         weights += layer.trainable_weights
     # dedupe weights
     weights = list(set(weights))
-    weights.sort(key=lambda x: x.name)
+    # TF variables have auto-generated the name, while Theano has auto-generated the auto_name variable. name in Theano is None
+    if weights:
+        if K.backend() == 'tensorflow':
+            weights.sort(key=lambda x: x.name)
+        elif K.backend() == 'theano':
+            weights.sort(key=lambda x: x.auto_name)
+        else:
+            raise Exception('Unknown backend: ' + str(K._BACKEND))
```

Comment:
K, I adjusted that accordingly.
Nevertheless I'm not sure, if that's the way to deal with it. If another backend would be used, you don't know of sorting by `name` would work, right?